eng/packages/http-client-csharp/generator/Azure.Generator/src/Visitors/DistributedTracingVisitor.cs

```diff
@@ -0,0 +1,264 @@
+            MethodProvider method)
+        {
+            if (method.BodyStatements != null)
+            {
+                PropertyProvider clientDiagnosticsProperty = method.EnclosingType.Properties
+                    .First(p => p.Name == ClientDiagnosticsPropertyName);
+                List<MethodBodyStatement> updatedFactoryMethodStatements = [];
+                foreach (var statement in method.BodyStatements.Flatten())
+                {
+                    if (TryUpdateSubClientFactoryMethodReturnStatement(
+                        statement,
+                        clientDiagnosticsProperty,
+                        out MethodBodyStatement? updatedStatement))
+                    {
+                        updatedFactoryMethodStatements.Add(updatedStatement);
+                    }
+                    else
+                    {
+                        updatedFactoryMethodStatements.Add(statement);
+                    }
+                }
+
+                method.Update(bodyStatements: updatedFactoryMethodStatements);
+            }
+        }
+
+        private static void UpdateProtocolMethodsWithDistributedTracing(MethodProvider method)
+        {
+            string scopeName = $"{method.EnclosingType.Name}.{method.Signature.Name}";
+            PropertyProvider clientDiagnosticsProperty = method.EnclosingType.Properties.First(p => p.Name == ClientDiagnosticsPropertyName);
+
+            // declare scope
+            var scopeDeclaration = UsingDeclare(
+                "scope",
+                typeof(DiagnosticScope),
+                clientDiagnosticsProperty.Invoke(nameof(ClientDiagnostics.CreateScope), [Literal(scopeName)], false, false),
+                out var scope);
+            // start scope
+            var scopeStart = scope.Invoke(nameof(DiagnosticScope.Start)).Terminate();
+            // wrap existing statements in try / catch
+            var tryStatement = new TryStatement
+                {
+                    method.BodyStatements ?? MethodBodyStatement.EmptyLine
+                };
+            var catchBlock = new CatchStatement(Declare("e", typeof(Exception), out var exception))
+                {
+                    scope.Invoke(nameof(DiagnosticScope.Failed), [exception]).Terminate(),
+                    Throw()
+                };
+            var tryCatchRequestBlock = new TryCatchFinallyStatement(tryStatement, catchBlock);
+            List<MethodBodyStatement> updatedBodyStatements = [scopeDeclaration, scopeStart, MethodBodyStatement.EmptyLine, tryCatchRequestBlock];
+
+            method.Update(bodyStatements: updatedBodyStatements);
+        }
+
+        private static void AddDistributedTracingProperty(ClientProvider client)
+        {
+            // Enable distributed tracing if the ClientDiagnostics property is not already present
+            if (client.CanonicalView.Properties.Any(p => p.Name == ClientDiagnosticsPropertyName))
```

Comment:
Not sure if we need this check as the visit method should only get invoked once.